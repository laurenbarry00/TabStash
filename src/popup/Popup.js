/*global chrome*/
import React, {Component} from 'react';

class Popup extends Component {

  constructor(props) {
    super(props);
    this.state = {
      alertText: ""
    }

    this.handleStash = this.handleStash.bind(this);
    this.handleApply = this.handleApply.bind(this);
    this.handleClear = this.handleClear.bind(this);
  }

  handleStash() {
    const _this = this;

    // query all tabs
    chrome.tabs.query({}, function(results) {
      let tabUrls = [];

      // get all open tabs' urls and indices
      results.forEach(function(tab) {
        tabUrls.push({
          url: tab.url,
          index: tab.index
        });
      });

      // store in chrome storage as json string
      //const openTabs = JSON.stringify(tabUrls);
      chrome.storage.local.set({tabStash: tabUrls});

      // update alert text w/ how many tabs were stored
      const alertString = tabUrls.length + " tabs stashed.";
      _this.setState({alertText: alertString});
    });
  }

  handleApply() {
    const _this = this;

    // retrieve stashed tabs from storage
    chrome.storage.local.get(["tabStash"], function(result) {
      // handle empty stash
      if (Object.keys(result).length == 0) {
        // display empty stash message
        const alertString = "Stash is empty.";
        _this.setState({alertText: alertString});
        return;
      }

      // get list of tabs
      const stashTabs = result.tabStash;
      stashTabs.forEach(function(tab) {
        // create new tab w/ index
        chrome.tabs.create({
          url: tab.url,
          index: tab.index
        });
      });
    })
  }

  handleClear() {
    // remove stash from storage
    chrome.storage.local.remove("tabStash");
    // update alert text to confirm stash clear
    this.setState({alertText: "Stash cleared."});
  }

  render() {
    const alertText = this.state.alertText;
    return (
      <div className="Popup">
        <h1 id="title">TabStash</h1>
        <p className="alert-text">{alertText}</p>
        <div className="buttons-container">
          <button id="stash-button" title="Stash current tabs" onClick={this.handleStash}>Stash</button>
          <button id="apply-button" title="Apply stashed tabs" onClick={this.handleApply}>Apply</button>
          <button id="clear-button" title="Clear stashed tabs" onClick={this.handleClear}>Clear</button>
        </div>
      </div>
    );
  }
}

export default Popup;
