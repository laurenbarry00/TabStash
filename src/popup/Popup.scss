.popup {
  display: flex;
  flex-direction: column;

  h1 {
    margin: 2%;
    padding: 10px;
    text-align: center;
    font-weight: normal;
  }

  img {
    height: auto;
    width: 64px;
    align-self: center;
  }

  .alert-text {
    align-self: center;
    height: 14px;
    display: block;
  }

  .error { color: red }
}

.main-buttons-container {
  display: flex;
  flex-direction: row;
  align-items: stretch;
  justify-content: center;

  button {
    padding: 5px 10px;
    margin: 0px 10px;

    @include button;
  }
}

#adv-menu-button {
  color: black;
  text-decoration: none;
  font-size: 16px;
  user-select: none;
  align-self: center;
  padding: 3px 10px;
  margin: 20px 10px;
  margin-bottom: 0px;

  &:active, &:hover {
    color: black;
    text-decoration: none;
    cursor: pointer;
    background-color: darken(white, 10%);
  }
}

.advanced-menu {
  display: flex;
  flex-direction: column;
  align-items: stretch;
  justify-content: center;
  margin: 10px 40px;

  button {
    margin: 5px;
    padding: 5px;
    @include button;

    &:first-child {
      margin-top: 0px;
    }
  }
}
