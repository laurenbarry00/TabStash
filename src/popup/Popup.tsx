import React from 'react';

function Popup() {
  return (
    <div className="Popup">
      <h1 id="title">TabStash</h1>
    </div>
  );
}

export default Popup;
