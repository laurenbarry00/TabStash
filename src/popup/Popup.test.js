import React from "react";
import Popup from "../popup/Popup";
import { mount, shallow, render } from "enzyme";

describe("Rendering" , () => {
  it("renders without crashing", () => {
    const component = shallow(<Popup />);
    expect(component).toMatchSnapshot();
  });

  it("renders the popup title", () => {
    const component = shallow(<Popup />);
    expect(component.find("#title").exists()).toBe(true);
    component.unmount();
  });

  it("renders the stash button correctly", () => {
    const component = shallow(<Popup />);
    expect(component.find("#stash-button").text()).toEqual("Stash");
  });

  it("renders the apply button correctly", () => {
    const component = shallow(<Popup />);
    expect(component.find("#apply-button").text()).toEqual("Apply");
  })

  it("renders the clear button correctly", () => {
    const component = shallow(<Popup />);
    expect(component.find("#clear-button").text()).toEqual("Clear");
  })
});

describe("Alert Text", () => {
  it("renders empty alert text", () => {
    const component = shallow(<Popup />);
    expect(component.find(".alert-text").exists()).toBe(true);
  })
})
