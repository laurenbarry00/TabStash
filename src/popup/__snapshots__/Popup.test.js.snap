// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders without crashing 1`] = `
<div
  className="Popup"
>
  <h1
    id="title"
  >
    TabStash
  </h1>
  <p
    className="alert-text"
  />
  <div
    className="buttons-container"
  >
    <button
      id="stash-button"
      onClick={[Function]}
      title="Stash current tabs"
    >
      Stash
    </button>
    <button
      id="apply-button"
      onClick={[Function]}
      title="Apply stashed tabs"
    >
      Apply
    </button>
    <button
      id="clear-button"
      onClick={[Function]}
      title="Clear stashed tabs"
    >
      Clear
    </button>
  </div>
</div>
`;
